### DISCUSSION TITLE
is it possible to make mqtt calls rather than http from outside the cluster ?

### DISCUSSION THREAD

**Discussion by Epado** [mia-platform/community#114] [Q&A] [answered]
I would like to make an mqtt call towards a service inside the cluster, is there something i should be warned about ? Is it possible ? Can you suggest a strategy ?

**Comment by fredmaggiowski** [comment, marked as answer]
MQTT runs over TCP so I think that there's no problem with that; do you see some technical details I'm missing?

**Reply by Epado** [reply]
Could you suggest me a way of debugging/undesrtanding how to make the call get through the api-gateway and to the actual broker ? I am not sure how to handle it!

**Reply by davidebianchi** [reply]
Have you tried exposing a normal endpoint? It depends on the project, what should expose?
We have not experience with mqtt, and we sure need more information about the final expected result to try to help you.

**Reply by Epado** [reply]
I have tried, but i guess i am missing some pieces on the lower layer. Il'll try and explain my goal;
I have a mqtt broker, which exposes mqtt over websockets. The structure of the project is the following; my microservice which runs the broker is inside a project (or namespace) which is called by a second project (or namespace) trough a proxy cross-project, my second project then has an api-gateway and exposes eveything to the outside  (Of course if the previously described architecture is not suited to the application with mqtt i can change it)
I tried and looked for 'how to expose a mqtt broker to the outside from a k8s cluster' and i found that i need to define an ingress and setup a http reverse proxy that points to my service to the port (9001) where the WebSocket Listener is listening but i don't know how to do it via console (again, i could go directly toward the project containing the mqtt broker if needed)

**Reply by fredmaggiowski** [reply]
if you need to connect with websocket from the outside than it seems like you just need to enable websockets on your API Gateway.

If that is the case here's the doc for [nginx](https://docs.mia-platform.eu/docs/development_suite/api-console/advanced-section/api-gateway/how-to#how-to-proxy-web-socket-connections) and [envoy](https://docs.mia-platform.eu/docs/development_suite/api-console/advanced-section/api-gateway-envoy/websockets).

**Reply by davidebianchi** [reply]
What type of api-gateway do you use? Do you want to expose through the gateway or directly from the mqtt broker? Are your project running in PaaS with traefik ingress?

**Reply by Epado** [reply]
Yes so far the project is running on PaaS with traefik ingress, api-gateway is the service based on Nginx version api-gateway:4.1.4. I'd say the easiest one at the moment, i think that passing though the gateway could increase latency but at the moment i cannot say how much, and if there is no need to bypass the gateway i'd prefer that

**Reply by Epado** [reply]
thanks for the info, about the doc for nginx, could you please explain better the following variables?
- $host : in this order of precedence, host name from the request line, or host name from the Host request header field, or the server name matching a request
- $proxy_add_x_forwarded_for : the X-Forwarded-For client request header field with the $remote_addr variable appended to it, separated by a comma
- $http_upgrade : the upgrade client request header

**Reply by davidebianchi** [reply]
This are variable already set inside our api-gateway based on nginx. The description is about how will be filled during runtime

**Reply by Epado** [reply]
i tried the following in `server-extension.conf`

```
location /mosquitto {
    proxy_pass http://mosquitto-broker;

    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header Host $host;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    # WebSocket support
    proxy_http_version 1.1;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection upgrade;
}
```

 where `/mosquitto` is an endpoint and `mosquitto-broker` is a proxy cross-project, do you think it's correct? i am trying to connect from outside but i am not able to, api-gateway is not logging, where should i look for debugging?

**Reply by Epado** [reply]
probably proxy_pass http://mosquitto-broker/; is wrong there is the hostname before, but still why is the api-gateway not logging, how can i know if the call is actually trying to enter the cluster?

**Reply by Epado** [reply]
Summing up it is possibile but the protocol needs to be mqtt over websockets. That being said you need 2 more points:

- configure the api-gateway to proxy websocket connections, which in my case was done writing the following in `server-extension.conf `:

```
location /mosquitto {
    proxy_pass http://mosquitto-broker;

    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header Host $host;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    # WebSocket support
    proxy_http_version 1.1;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection upgrade;
}
```
- enable tls on the client side, and use the standard port for https connections 443 rather than the standard one for the mqtt secure connections.

Thanks again for the support, Cheers!
